Uncertainty funcs dxu

danielxu9393 opened the pull request:
## Description
Created support for conformal prediction, adaptive conformal prediction, and conformal regression.


## Example / Current workflow

User enters command line query to predict.py with TAP arguments. Calls function make_predictions in make_predictions.py. make_predictions loads model, initializes calibrator using build_uncertainty_calibrator, and calls predict_and_save. predict_and_save initializes uncertainty estimator, extracts results and outputs into csv. When uncertainty estimator is initialized, the estimator will take the loaded model and create the uncertainty_predictor and apply the calibrator to the uncalibrated outputs.

For conformal quantile regression, an additional training workflow is implemented. The model is changed to have two heads which output the $\alpha/2$ and $1-\alpha/2$ quantiles.



Usage described below. All queries made from the /conformal folder. Train examples in the .py files in the /conformal folder. (from /conformal folder):

Conformal Adaptive:
Use the --alpha parameter to change the value of alpha, default is 0.1
For basic conformal, use --calibration_method conformal

python3 ../predict.py --test_path data/tox21.csv --checkpoint_dir tox21_checkpoints2 --preds_path tox21_preds_conformal_adaptive.csv --calibration_method conformal_adaptive --calibration_path data/tox21small.csv

python3 ../predict.py --test_path data/tox21.csv --checkpoint_dir tox21_checkpoints2 --preds_path tox21_preds_conformal_adaptive_alpha05.csv --calibration_method conformal_adaptive --calibration_path data/tox21small.csv --alpha 0.5

Conformal Regression:
Again, use the --alpha parameter to change the value of alpha. Default is 0.1

python3 ../predict.py --test_path data/delaney.csv --checkpoint_dir delaney_checkpoints2 --preds_path delaney_preds_conformal.csv --calibration_method conformal_regression --calibration_path data/delaneysmall.csv

Conformal Interval Regression:

First train the model using the --loss_function quantile_interval and --alpha <float> parameters. This creates a model that outputs the $\alpha/2$ and $1-\alpha/2$ quantiles. Inference can then be done very similar to the above examples.

python3 ../predict.py --test_path data/delaney.csv --checkpoint_dir delaney_checkpoints_quantile_interval05 --preds_path delaney_preds_conformal_quantile_interval05.csv --calibration_method conformal_quantile_regression --alpha 0.1 --calibration_path data/delaneysmall.csv

danielxu9393 committed:
Conformal Predictor

danielxu9393 committed:
Conformal Calibrator with Hardcoded qhat

danielxu9393 committed:
calculated qhat, hardcoded alpha

danielxu9393 committed:
Conformal basic complete

danielxu9393 committed:
First Version Conformal

danielxu9393 committed:
updated readme to include conformal

danielxu9393 committed:
cleaned up code

danielxu9393 committed:
typo in readme

danielxu9393 committed:
fixed conformal_regression output

ajfisch commented on README.md:
```diff
@@ -374,9 +374,14 @@ Uncertainty predictions may be calibrated to improve their performance on new pr
 * `tscaling` Similar to zscaling. Assumes that the errors are normally distributed, but accounts for the ensemble size and uncertainty in the sample variance by using a sample-size reduced t-distribution in the negative log likelihood. Works best when errors are mostly due to variability between model instances and not dataset noise or model bias.
 * `zelikman_interval` Assumes that the error distribution is the same for each prediction but scaled by the uncalibrated standard deviation for each. Multiplies the uncalibrated standard deviation by a factor necessary to cover the specified interval of the calibration set. Does not assume a Gaussian distribution. Intended for use with intervals but can return a stdev as well. (https://arxiv.org/ ...
 * `mve_weighting` For use with ensembles of models trained with mve or evidential loss function. Uses a weighted average of the predicted variances to achieve a minimum negative log likelihood of predictions. (https://doi.org/10.1186/s13321-021-00551-x)
+* `conformal_regression` Generates an interval of fixed size for each prediction such that the actual value has probability 1-alpha of falling in the set. The user can control the desired error rate using the parameter `--alpha <float>` which is set by default to 0.1. Use any uncertainty method. (https://arxiv.org/abs/2107.07511)
```
falling in the set --> falling in the interval

ajfisch commented on chemprop/args.py:
```diff
@@ -755,9 +755,11 @@ class PredictArgs(CommonArgs):
         'classification',
         'dropout',
         'spectra_roundrobin',
+        'conformal',#added new possibility
```
remove comment

ajfisch commented on chemprop/train/make_predictions.py:
```diff
@@ -238,16 +239,18 @@ def predict_and_save(
             ]
-            num_tasks = num_tasks * args.multiclass_num_classes
-        if args.uncertainty_method == "spectra_roundrobin":
+            num_tasks = num_tasks * args.multiclass_num_classes#num_tasks is like number of classes, num_unc_tasks is same but for uncertaint pred output
```
remove comments

ajfisch commented on chemprop/train/make_predictions.py:
```diff
@@ -238,16 +239,18 @@ def predict_and_save(
         for full_index, datapoint in enumerate(full_data):
             valid_index = full_to_valid_indices.get(full_index, None)
-            d_preds = (
+            d_preds = (#taking preds, unc and filtering out invalid datapoints
```
Seems like a good comment to leave in, but format it properly:
```
# d_preds contains the raw predictions, d_unc contains uncertainty values.
# Both filter out invalid datapoints.
d_preds = (
...
)
d_unc = (
...
)
```

ajfisch commented on chemprop/train/make_predictions.py:
```diff
@@ -285,15 +288,31 @@ def predict_and_save(
             if args.uncertainty_method == "spectra_roundrobin":
                 unc_names = [estimator.label]
+            elif args.calibration_method == "conformal_regression":
+                unc_names = [task_names[0] + "_conformal_regression_lower_bound", task_names[0] + "_conformal_regression_upper_bound"]
```
We are assuming that `task_names` is always a list of length 1? Is this true?

ajfisch commented on chemprop/train/make_predictions.py:
```diff
@@ -285,15 +288,31 @@ def predict_and_save(
                 unc_names = [name + f"_{estimator.label}" for name in task_names]
-
+            
+            """
```
Seems like you want to delete this rather than leave as a comment? It would be good to instead make a comment as to why `task_names, d_preds` and `unk_names, d_unc` can be different shapes.

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +620,223 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
         return uncal_preds.tolist(), cal_preds.tolist()
 
 
+class ConformalCalibrator(UncertaintyCalibrator):
```
Apply pylint to this. Specifically, we want no more than 2 blank lines within the body of a class. 2 blank lines between classes.

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +620,223 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+            )
+
+    def calibrate(self):
+        uncal_preds = np.array(
```
This can be one line. Also, what type do we expect `get_uncal_preds()` to return? Can it definitely be cast to a numpy array? Note that pytorch tensors should be cast to numpy arrays using `tensor.cpu().numpy()`. It's way faster.

danielxu9393 replied:
It seems like the way they implemented it, the output of predictor is always cast to a list.

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +620,223 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+        calibration_set = np.array(calibration_set, dtype=int)
+
+        def s_basic(i, j, uncal_preds):#x,y are the indices
+                return -uncal_preds[i][j]
```
4 spaces indent

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +620,223 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+        (self.num_data, self.num_tasks) = targets.shape
+        calibration_set = []
+
+        for i in range(self.num_data):
```
Things like this can and should always be done neatly in vectorized numpy ops.

If you don't know what this is, you can read up on it online. E.g.,  https://www.pythonlikeyoumeanit.com/Module3_IntroducingNumpy/VectorizedOperations.html.

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +620,223 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+            qhat = np.quantile(calibration_scores, 1-alpha)
+
+            return qhat
+
```
One blank line

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +620,223 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+        s = s_basic
+        self.qhat = calculate_qhat(calibration_set, uncal_preds, s, self.alpha)
+
+        
```
one blank line

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +620,223 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+        
+        cal_preds=[[int(s(i, j, uncal_preds) <= self.qhat) for j in range(K)] for i in range(N)]
+        return uncal_preds.tolist(), cal_preds
+
```
2 blank lines

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +620,223 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+        cal_preds=[[int(s(i, j, uncal_preds) <= self.qhat) for j in range(K)] for i in range(N)]
+        return uncal_preds.tolist(), cal_preds
+
+class ConformalAdaptiveCalibrator(UncertaintyCalibrator):
```
Can we make this class inherit from `ConformalCalibrator`? This has too much copy paste.

Instead of redefining `s` each time to be `s_basic` or `s_adaptive`, lets make `s` a proper function. For example, in the case of the standard conformal score:

```python
def nonconformity_score(self, i, j):
    return -self.uncalibrated_predictions[i, j]
```
(But we can also make all of these functions vectorized, so maybe don't actually call for [i, j] in a for loop).
Perhaps let's also store `uncal_preds` as a class variable to make the call cleaner.

danielxu9393 replied:
The reason I had the uncal_preds as an input to s is because there are different uncal_preds depending on if it is calibrating or generating outputs. So I thought it would be easier to have just one s function.

danielxu9393 committed:
Added quantile_loss with hardcoded quantile

danielxu9393 committed:
Support for quantile parameter in command line

danielxu9393 committed:
Calibration method conformal_quantile_regression works

danielxu9393 committed:
Quantile Interval Multihead training

danielxu9393 committed:
Finished conformal quantile regression

danielxu9393 committed:
updated readme for conformal quantile training and inference

danielxu9393 committed:
Cleaned up conformal folder, updated README

ajfisch commented on README.md:
```diff
@@ -197,6 +197,7 @@ The loss functions available for training are dependent on the selected dataset
 * **Classification.** binary_cross_entropy (default), mcc (a soft version of Matthews Correlation Coefficient), dirichlet (a.k.a. evidential classification)
 * **Multiclass.** cross_entropy (default), mcc (a soft version of Matthews Correlation Coefficient)
 * **Spectra.** sid (default, spectral information divergence), wasserstein (First-order Wasserstein distance a.k.a. earthmover's distance.)
+* **Quantile Loss.** quantile_interval (specify margins with `--alpha <float>`.)
```
Make this a bit more descriptive (i.e., the pinball loss).

ajfisch commented on chemprop/models/model.py:
```diff
@@ -115,11 +118,24 @@ def create_ffn(self, args: TrainArgs) -> None:
+                    nn.Linear(args.ffn_hidden_size, 1),
+                ])
+            else:
+                ffn.extend([
```
You don't need this. By specifying the output size as 2, you have already effective created two different linear heads, and this should be sufficient. 

Notice that the final output will be computed as:
```
y = Wx + b
```
where y is `2x1`, W is `2xd`, x is `dx1`, and b is `2x1`. This is the same as computing
```
y_1 = w_1'x + b_1
y_2 = w_2'x + b_2
```
but just in matrix form.

danielxu9393 replied:
Your right, this is legacy from when I was trying to save into 2 different checkpoints.

ajfisch commented on chemprop/models/model.py:
```diff
@@ -216,5 +232,11 @@ def forward(self,
             output = torch.cat([means, lambdas, alphas, betas], dim=1)
         if self.loss_function == 'dirichlet':
             output = nn.functional.softplus(output) + 1
+        if self.loss_function == 'quantile_interval':
```
Again, we don't need this. Also, this function expects a torch tensor output.

ajfisch commented on chemprop/train/loss_functions.py:
```diff
@@ -240,6 +242,19 @@ def normal_mve(pred_values, targets):
 
+def quantile_loss(pred_values, targets, quantile=0.5):
+    """
+    Use the negative log likelihood function of a normal distribution as a loss function used for making
```
Update comment.

ajfisch commented on chemprop/train/loss_functions.py:
```diff
@@ -240,6 +242,19 @@ def normal_mve(pred_values, targets):
         2 * pred_var
     )
 
+def quantile_loss(pred_values, targets, quantile=0.5):
```
If you are training multiple quantiles simultaneously, why not make this loss parallel in the number of outputs as well.

Assume inputs are of size
```
pred_values = [batch_size, num_outputs]
targets = [batch_size]
quantiles = [num_outputs]
```
Then you can simply expand targets to be of size [batch_size, num_outputs] as well, and return a loss of size [num_outputs].

ajfisch commented on chemprop/train/make_predictions.py:
```diff
@@ -403,13 +426,21 @@ def make_predictions(
+        elif args.calibration_method == 'conformal_regression':
+            args.uncertainty_method = None
+        elif args.calibration_method == 'conformal_quantile_regression':
+            args.uncertainty_method = 'conformal_quantile_regression'
```
Why is uncertainty_method None in one case but not the other?

ajfisch commented on chemprop/train/predict.py:
```diff
@@ -37,7 +37,12 @@ def activate_dropout_(model):
         model.apply(activate_dropout_)
 
-    preds = []
+    if model.loss_function == 'quantile_interval':
```
Again, once the output is 2d, just leave it. Don't need these hacks here.

ajfisch commented on chemprop/train/predict.py:
```diff
@@ -60,7 +65,11 @@ def activate_dropout_(model):
                 bond_features_batch,
             )
 
-        batch_preds = batch_preds.data.cpu().numpy()
```
Same. Leaving the outputs as 2 dimensional will render the following hack unnecessary.

FYI `tensor.data.cpu().numpy()` is used to get the raw underlying data of a tensor that might have autograd enabled, transfers the tensor to the cpu if it is on the gpu, and casts it a numpy array.

ajfisch commented on chemprop/train/train.py:
```diff
@@ -106,6 +110,12 @@ def train(model: MoleculeModel,
             loss = loss_func(preds, targets, args.evidential_regularization) * target_weights * data_weights * mask
+        elif args.loss_function == 'quantile':
+            loss = loss_func(preds, targets, args.quantile) * target_weights * data_weights * mask
+        elif args.loss_function == 'quantile_interval':
```
Here you'd either split the last two dimensions of the output tensor and call loss_func twice as you do here, or make the quantile loss handle multiple target quantiles at the same time. If you did that, then you'd just call something like:

```python
loss = loss_func(preds, targets, preds.new([1 - args.alpha / 2, args.alpha / 2]))
```

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +622,285 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+        uncal_preds = np.array(
+            self.calibration_predictor.get_uncal_preds()
+        )  # shape(data, tasks, num_classes)
+        targets = np.array(self.calibration_data.targets(), dtype=float)  # shape(data, tasks)
```
I think this is actually intended to be multi-label classification? If the shape is `[num_examples, num_tasks]` then it's really performing predictions of {0, 1} for num_tasks different targets. In this case, you need to do multi-label conformal (https://arxiv.org/abs/2004.10181). 

Let's talk about this later.

ajfisch reviewed (changes requested):
Hi Daniel!

Thanks for your great work on this so far. I left a number of initial comments. There are a bunch of things that should be iterated on, so let's take a stab at fixing those and I can re-review after they are addressed for more comments.

A number of high level comments:
- Let's take the time to get into the good habit of nicely formatting our code. Read up on Google's python style guide. In particular spacing and comments.
- Use pytorch and numpy inbuilt functions and vectorized operations wherever possible. This is faster, cleaner, and easier to follow than rolling your own for loops.
- We have to talk about multi-class vs. multi-label classification. I think that the format you have used for classification is actually the one for binary multi-label prediction (the ones that have targets of size [num_examples, num_tasks]). There is a very simple fix we can do, but we should discuss via Zoom.

danielxu9393 committed:
Fixed spacing and comments

danielxu9393 committed:
refactored quantile_interval training to support multiple tasks and got rid of dictionary formatting

danielxu9393 committed:
refactored code to change task_names to get rid of redundant code

danielxu9393 committed:
Updated inference workflow for conformal_quantile to match changes

danielxu9393 committed:
multilabel conformal complete

danielxu9393 committed:
fixed multiclass conformal

ajfisch commented on chemprop/args.py:
```diff
@@ -766,6 +779,10 @@ class PredictArgs(CommonArgs):
     """Location to save the results of uncertainty evaluations."""
     uncertainty_dropout_p: float = 0.1
     """The probability to use for Monte Carlo dropout uncertainty estimation."""
+
```
nit: extra spaces above and below

ajfisch commented on chemprop/models/model.py:
```diff
@@ -121,6 +121,7 @@ def create_ffn(self, args: TrainArgs) -> None:
                 nn.Linear(args.ffn_hidden_size, self.output_size),
             ])
 
+
```
nit: extra space

ajfisch commented on chemprop/train/make_predictions.py:
```diff
@@ -282,18 +289,36 @@ def predict_and_save(
+            elif args.calibration_method == "conformal_regression":
+                unc_names = [task_name + "_conformal_regression_lower_bound" for task_name in task_names] \
+                + [task_name + "_conformal_regression_upper_bound" for task_name in task_names]
+            elif args.calibration_method == "conformal_quantile_regression":
```
You can share the same names (lower bound/upper bound). With the understanding that one is adaptive, and the other is fixed size.

danielxu9393 replied:
So should the names be something like task_name + "lower_bound" and get rid of conformal_regression or conformal_quantile_regression?

ajfisch replied:
Yeah, sgtm!

ajfisch commented on chemprop/train/make_predictions.py:
```diff
@@ -403,13 +428,22 @@ def make_predictions(
     if args.uncertainty_method is None and (args.calibration_method is not None or args.evaluation_methods is not None):
         if args.dataset_type in ['classification', 'multiclass']:
             args.uncertainty_method = 'classification'
+        elif args.calibration_method == 'conformal_regression':
```
combine these two checks (`elif args.calibration_method in ["conformal_regression", "conformal_quantile_regression"]`)

ajfisch commented on chemprop/train/make_predictions.py:
```diff
@@ -403,13 +428,22 @@ def make_predictions(
             raise ValueError('Cannot calibrate or evaluate uncertainty without selection of an uncertainty method.')
 
 
+
```
nit: extra space

ajfisch commented on chemprop/train/predict.py:
```diff
@@ -85,6 +85,7 @@ def activate_dropout_(model):
         # Inverse scale if regression
         if scaler is not None:
             batch_preds = scaler.inverse_transform(batch_preds)
+
```
nit: extra space

ajfisch commented on chemprop/train/train.py:
```diff
@@ -106,6 +105,10 @@ def train(model: MoleculeModel,
+        elif args.loss_function == 'quantile':
+            loss = loss_func(preds, targets, args.quantile) * target_weights * data_weights * mask
+        elif args.loss_function == 'quantile_interval':
+            loss = loss_func(preds, targets, torch.tensor([args.quantiles])) * target_weights * data_weights * mask
```
It's important that you specify the device (cuda vs. cpu) when making a new tensor like this. You can use 

```python
torch.tensor([args.quantiles]).to(torch_device)
```
(see line 70)

Or the way I wrote before using `preds.new`.

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -650,6 +890,12 @@ def build_uncertainty_calibrator(
         "zelikman_interval": ZelikmanCalibrator,
         "mve_weighting": MVEWeightingCalibrator,
         "platt": PlattCalibrator,
+        "conformal": ConformalMultilabelCalibrator
```
Can this be one line?

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +621,239 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+    def nonconformity_score(self, i, j, uncal_preds):
+        K=uncal_preds.shape[1]
+        result = 0
+        for Y in range(K):
```
Can we numpify these slow for loops?

For example, here we would write this:

```python
ge_indices = uncal_preds[i] >= uncal_preds[i, j]
result = uncal_preds[i, ge_indices].sum()
```
EDIT: see note on line 642

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +621,239 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+
+        calibration_set = np.array(calibration_set, dtype=int)
+
+        def calculate_qhat(calibration_set, uncal_preds, nonconformity_score, alpha):
```
We don't need this to be a separate function, the logic is only used once.

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +621,239 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+        (self.num_data, self.num_tasks) = targets.shape
+        calibration_set=[]
+
+        for i in np.arange(self.num_data):
```
We don't need this step. It's slow, and we can use `targets` (array of size [num_examples] with class indices) directly.

danielxu9393 committed:
multiclass multiple tasks

danielxu9393 committed:
bonferroni correction and multi task regression

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +621,265 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+    def nonconformity_score(self, i, j, task_id, uncal_preds):#x,y are the indices
+        return -uncal_preds[i][task_id][j]
+
+    def calibrate(self):
```
There is a more concise and faster way to write these functions, while avoiding long for loops over all the data. I have written some quick but untested code that does what you want to do this way.

```python
@staticmethod
def nonconformity_scores(preds):
    """Fixed per class. Example is for adaptive.                                                                                                                                                                                                                              
                                                                                                                                                                                                                                                                              
    Args:                                                                                                                                                                                                                                                                     
        preds: [num_examples, num_tasks, num_classes]                                                                                                                                                                                                                         
                                                                                                                                                                                                                                                                              
    Returns:                                                                                                                                                                                                                                                                  
        scores: [num_examples, num_tasks, num_classes]                                                                                                                                                                                                                        
    """
    # Sort from high to low.                                                                                                                                                                                                                                                  
    sort_inds = np.argsort(-preds, axis=-1)

    # Sort, and take cumulative sums.                                                                                                                                                                                                                                         
    sorted_scores = preds[sort_inds].cumsum(axis=-1)

    # Map back to original order.                                                                                                                                                                                                                                             
    unsort_inds = np.argsort(sort_inds, axis=-1)
    unsorted_scores = sorted_scores[unsorted_inds]

    return unsorted_scores

def calibrate(self):
    # shape(num_examples, num_tasks, num_classes)                                                                                                                                                                                                                             
    uncal_preds = np.array(self.calibration_predictor.get_uncal_preds())

    # shape(num_examples, num_tasks)                                                                                                                                                                                                                                          
    targets = np.array(self.calibration_data.targets(), dtype=int)

    # Keep mask of invalid targets (missing data). Replace Nans with 0.                                                                                                                                                                                                       
    mask = np.isnan(targets)
    targets = np.nan_to_num(targets, copy=True, nan=0, posinf=None, neginf=None)

    # Compute all non-conformity scores for all examples/tasks/classes.                                                                                                                                                                                                       
    all_scores = self.nonconformity_scores(uncal_preds)

    # Take scores of true targets only.                                                                                                                                                                                                                                       
    target_scores = np.take_along_axis(all_scores, np.expand_dims(targets, -1), -1).squeeze(-1)

    # For every task, compute calibration set quantile.                                                                                                                                                                                                                       
    self.qhats = []
    num_tasks = targets.shape[1]
    for task_id in range(num_tasks):
        # Get all vaid non-conformity scores for this task.                                                                                                                                                                                                                   
        task_scores = target_scores[mask[:, task_id], task_id].reshape(-1)

        # Pad with +inf.                                                                                                                                                                                                                                                      
        task_scores = np.pad(task_scores, (0, 1), "constant", constant_values=np.Inf)

        # Compute empirical quantile.                                                                                                                                                                                                                                         
        qhat = np.quantile(task_scores, 1 - self.alpha / num_tasks, interpolation="upper")
        qhats.append(qhat)


def apply_calibration(self, uncal_predictor):
    # shape(num_examples, num_tasks, num_classes)                                                                                                                                                                                                                             
    uncal_preds = np.array(uncal_predictor.get_uncal_preds())  # shape(data, task)                                                                                                                                                                                            

    # Compute scores.
    uncal_scores = self.nonconformity_scores(uncal_preds)

    # 0/1 for if candidate class y is included in output set.                                                                                                                                                                                                                 
    # shape(num_examples, num_tasks, num_classes)                                                                                                                                                                                                                             
    cal_preds = np.zeros_like(uncal_preds)
    for task_id, qhat in enumerate(self.qhats):
        cal_preds[:, task_id, :] = uncal_scores[:, task_id, :] <= qhat

    return uncal_preds.tolist(), cal_preds.tolist()
```

ajfisch reviewed (changes requested):
Looking better. See main point on the multiclass.

ajfisch reviewed (commented):
Please also remove log and binary files

danielxu9393 committed:
refactored multiclass conformal

danielxu9393 committed:
removed comments from multiclass conformal

danielxu9393 committed:
refactored multilabel and modularized/refactored regression

ajfisch commented on conformal/python_test.py:
```diff
@@ -0,0 +1,67 @@
+sorted = np.take_along_axis(x, sort_ind, axis=2)
+print(sorted)
+
+#print(np.argsort(x, axis=0)) # idk what this does? sorts each pipe but along axis=0??? Viewing as nd thingy?
```
See https://numpy.org/doc/stable/reference/generated/numpy.argsort.html. argsort(x, axis=0) would sort a matrix x down its columns.

So if x is 
```
[[1, 5, 3],
  [7, 2, 1],
  [2, 0, 6]]
```
argsort would return 
```
[[0, 2, 1],
  [2, 1, 0],
  [1, 0, 2]]
```

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +621,253 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+        sort_inds = np.argsort(-uncal_preds, axis=2)
+
+        # Sort and calculate scores
+        sorted_preds = np.take_along_axis(uncal_preds, sort_inds, axis=2)
```
nice catch fixing this from my previous note

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -615,6 +621,253 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
+                "Conformal is only implemented for classification dataset types."
+            )
+
+    def calibrate(self):
```
Ah, ok, I think you had a slight conceptual mistake here?

When computing
```python
self.tout_hat = np.quantile(calibration_scores_out, self.alpha / 2)
self.tin_hat = np.quantile(calibration_scores_in, 1 - self.alpha / 2)
```
you are trying to directly replace Eq. 15 of alg 3 with a quantile that is independent of x? You can only do this if tout_hat/tin_hat are learned (in this case, merely computed) on separate data (note the use of I_2 vs I_3 in Alg. 3). 

Instead what I was suggesting (to avoid needing to both estimate tout/tin and the associated data for this step), was to just use a much simpler (and, yes, less efficient) estimator:

```python
# Assume that lower nonconformity scores are better. i.e., the lower s_k(x), the more likely y_k = 1.

# Under the null that the new y_k is 0, s_k(x) will be greater than the alpha/2 quantile of min nonconf. scores when y_k is 0, at least 1 - alpha/2 fraction of the time.
# Flip scores to be negative to compute quantile, but this should roughly be the same as
# if s_k(x) > Quantile(alpha/2, min_k s_k(x) : Y_k = 0 \cup {0})
if -s_k(x) ≤ Quantile(1 - alpha/2, -min_k s_k(X) : Y_k = 0 \cup {inf}):
    C_in(k) = 0
else:
    C_in(k) = 1

# Under the null that the new y_k is 1, s_k(x) will be less than the 1-alpha/2 quantile of max nonconf. scores when y_k is 1, at least 1 - alpha/2 fraction of the time.
if s_k(x) ≤ Quantile(1 - alpha/2, max_k s_k(x) : y_k = 1 \cup {inf}):
    C_out(k) = 1
else:
    C_out(k) = 0
```
You can double check me on this. 


To do this efficiently in numpy... follow like this (note use of masking to ignore scores that are not meeting y_k = 0. Similar for y_k = 1.
```
has_zeros = np.sum(targets == 0, axis=1) > 0
inds_zero = targets[has_zeros] == 0  # Dont forget about ignoring the mask Nans too
... scores = compute_scores ... # this will be just -uncal_preds, but might want something else sometime later.
masked_scores = scores * (1 - inds_zero) + np.Inf * inds_zeros
min_zeros = np.min(masked_scores, axis=1)
... compute quantile...
```

danielxu9393 replied:
I see, so essentially just ignoring the qhat step and replacing tin, tout with tin_hat and tout_hat

danielxu9393 committed:
fixed and numpified multilabel

danielxu9393 committed:
small update to multilabel

test plan for regression done

removed extraneous files

added test_plan folder

removed test_plan/data folder

test_plan folder

danielxu9393 committed:
small update to multilabel

danielxu9393 committed:
test plan for regression done

danielxu9393 committed:
removed extraneous files

danielxu9393 committed:
test_plan

danielxu9393 committed:
some merge issue?

danielxu9393 committed:
test plan complete

danielxu9393 committed:
cleaned print statements

ajfisch commented on chemprop/uncertainty/uncertainty_calibrator.py:
```diff
@@ -549,7 +556,6 @@ def apply_calibration(self, uncal_predictor: UncertaintyPredictor):
         cal_preds = np.transpose(transpose_cal_preds)
         return uncal_preds.tolist(), cal_preds.tolist()
 
-
```
Don't remove this space?

ajfisch reviewed (approved):
Hi Daniel, this looks good!

Remove all of the files that are not code and relevant to this PR. So all of the output data files, tfrecords, etc. There should be like 5 files changed, not 270 :).

danielxu9393 committed:
removing test_plan

danielxu9393 committed:
removed conformal folder

danielxu9393 committed:
cleaned up
